model/flow/version_beacon.go

```diff
@@ -60,10 +70,78 @@ func (v *VersionBeacon) EqualTo(other *VersionBeacon) bool {
 		if v.BlockHeight != other.BlockHeight {
 			return false
 		}
-		if semver.Compare(v.Version, other.Version) != 0 {
+
+		v1, err := v.Semver()
+		if err != nil {
+			return false
+		}
+		v2, err := other.Semver()
+		if err != nil {
+			return false
+		}
+		if !v1.Equal(*v2) {
 			return false
 		}
 	}
 
 	return true
 }
+
+// Validate validates the internal structure of a flow.VersionBeacon.
+// An error with an appropriate message is returned
+// if any validation fails.
+func (v *VersionBeacon) Validate() error {
+	eventError := func(format string, args ...interface{}) error {
+		args = append([]interface{}{v.Sequence}, args...)
+		return fmt.Errorf(
+			"version beacon (sequence=%d) error: "+format,
+			args...,
+		)
+	}
+
+	if len(v.VersionBoundaries) == 0 {
+		return eventError("required version boundaries empty")
+	}
+
+	var previousHeight uint64
+	var previousVersion *semver.Version
+	for i, boundary := range v.VersionBoundaries {
+		version, err := boundary.Semver()
+		if err != nil {
+			return eventError(
+				"invalid semver %s for version boundary (height=%d) (index=%d): %w",
+				boundary.Version,
+				boundary.BlockHeight,
+				i,
+				err,
+			)
+		}
+
+		if i != 0 && previousHeight >= boundary.BlockHeight {
```

Comment:
Is this event from smart contract? In that case, the smart contract must have validated it? 

I guess, we are doing sanity check, just to be safe, right?